lib/Importer/Caffe2ModelLoader.cpp

```diff
@@ -238,145 +274,248 @@ bool Caffe2ModelLoader::hasMultidirectionalBroadcast(
-    auto outSz =
-        calculateConvPoolOutputDims(idim.h, idim.w, kernels, strides, pads);
-    std::array<size_t, 4> outDims = {
-        {idim.n, outSz.first, outSz.second, depth}};
+  // Load the inputs:
+  std::vector<unsigned_t> strides;
+  ASSIGN_VALUE_OR_RETURN_ERR(strides, getSizeHW(dict, "stride", 1));
+  std::vector<unsigned_t> pads;
+  ASSIGN_VALUE_OR_RETURN_ERR(pads, getPads(dict));
+  std::vector<unsigned_t> kernels;
+  ASSIGN_VALUE_OR_RETURN_ERR(kernels, getSizeHW(dict, "kernel", 0));
+  unsigned_t group = 1;
+  if (dict.count("group")) {
+    ASSIGN_VALUE_OR_RETURN_ERR(group, loadInt(dict["group"]));
+  }
+  std::string order = "NCHW";
+  if (dict.count("order")) {
+    ASSIGN_VALUE_OR_RETURN_ERR(order, loadStr(dict["order"]));
+  }
+  bool quantizeGroupwise = false;
+  if (dict.count("quantize_groupwise")) {
+    ASSIGN_VALUE_OR_RETURN_ERR(quantizeGroupwise,
+                               loadInt(dict["quantize_groupwise"]));
+  }
 
-    TypeRef outTy;
-    Constant *filter = W;
-    Constant *bias;
-    if (typeName == "Conv" || typeName == "ConvRelu") {
-      // Construct the Bias field.
-      Constant *biasConstant = nullptr;
-
-      // Check if we have a serialized bias.
-      if (op.input_size() > 2) {
-        const auto &biasName = op.input(2);
-        biasConstant = getConstantByNameOrNull(biasName);
-      }
+  // Group quantization only applies if there is more than one group.
+  quantizeGroupwise &= group > 1;
 
-      // If no serialized bias was found then create a zero bias.
-      if (!biasConstant) {
-        Tensor biasTensor(ElemKind::FloatTy, {depth});
-        biasTensor.zero();
-        biasConstant = G_.getParent()->createConstant("conv.bias", biasTensor);
-      }
+  NodeValue in;
+  ASSIGN_VALUE_OR_RETURN_ERR(in, getNodeValueByName(op.input(0)));
 
-      outTy = G_.getParent()->uniqueType(ElemKind::FloatTy, outDims);
-      bias = biasConstant;
-    } else {
-      RETURN_ERR_IF_NOT(dict.count("Y_zero_point"),
-                        "missing zero point for quantized output type");
-      RETURN_ERR_IF_NOT(dict.count("Y_scale"),
-                        "missing Y_scale for quantized output type");
-      // Construct the Bias field.
-      Constant *biasConstant = nullptr;
+  Constant *w;
+  ASSIGN_VALUE_OR_RETURN_ERR(w, getConstantByName(op.input(1)));
 
-      // Check if we have a serialized bias vector.
-      if (op.input_size() > 2) {
-        const auto &biasName = op.input(2);
-        biasConstant = getConstantByNameOrNull(biasName);
-      }
+  // Transpose the weights to the right format. Glow expects to read the
+  // weights in the format CRSK.
+  // C - output_depth, R - filter_height, S - filter_width, K - input_depth.
+  // For Caffe2 "Int8Conv" and "Int8ConvRelu", the weights always follows the
+  // "order" arg.
+  if (order != "NHWC") {
+    Tensor wT;
+    w->getPayload().transpose(&wT, NCHW2NHWC);
+    w = G_.getParent()->createConstant(w->getName(), std::move(wT));
```

Comment:
nit: I see this is copied over from the previous impl, and it's probably done elsewhere in the loaders too -- but I don't think we should be doing these sorts of transformations/duplicate Constant creation here. We should just add a Transpose node to the weight and let the Graph optimizer handle merging the Transpose in.

Reply:
Yeah it's good point, I think this wasn't so straight forward before we loaded things directly into `Constant`s instead of into `Tensor`s but now that we do it seems kind of weird to do it this way.

Reply:
Can we make this change in a follow up diff? I'd prefer not to increase the number of changes here more.